backend/src/db/providers/clientCourseProvider.ts

```diff
@@ -9,4 +16,24 @@ const getClientCoursesProvider = async () => {
   }
 };
 
-export { getClientCoursesProvider };
+const applyCourseProvider = async (courseId: string, userId: string) => {
+  const materialsCount = await CourseModel.aggregate([
+    { $match: { _id: new mongoose.Types.ObjectId(courseId) } },
+    {
+      $project: {
+        total: { $size: '$materials' },
+      },
+    },
+  ]);
+  const progressDto = generateProgressDto(materialsCount[INITIAL_INDX].total);
+  const applyedCourse = await ClientCourseModel.create({
+    course: courseId,
+    status: 'approved',
+    currentStage: 1,
+    progress: progressDto,
+  });
+  await UserModel.updateOne({ _id: userId }, { $push: { courses: applyedCourse.course } });
+  return applyedCourse;
```

Comment:
It should be transaction!
It should be inside controller, db providers can only execute 1-2 operations with DB, but all the logic should be inside controller